[MemoryLocation] Teach MemoryLocation about llvm.experimental.memset.pattern

asb opened the pull request:
Relates to (but isn't dependent on) #120420.

This allows alias analysis o the intrinsic of the same quality as for the libcall, which we want in order to move LoopIdiomRecognize over to selecting the intrinsic.

asb committed:
[test] Pre-commit llvm.experimental.memset.pattern tests prior to MemoryLocation changes

The next patch will teach MemoryLocation about
llvm.experimental.memset.pattern.

asb committed:
[MemoryLocation] Teach MemoryLocation about llvm.experimental.memset.pattern

nikic commented on llvm/lib/Analysis/MemoryLocation.cpp:
```diff
@@ -183,6 +183,18 @@ MemoryLocation MemoryLocation::getForArgument(const CallBase *Call,
+            Arg,
+            LocationSize::precise(
+                LenCI->getZExtValue() *
+                DL.getTypeStoreSize(II->getArgOperand(1)->getType())),
```
```suggestion
                DL.getTypeAllocSize(II->getArgOperand(1)->getType())),
```

preames commented:
> Stacks on top of https://github.com/llvm/llvm-project/pull/120420.

I don't see any obvious dependency on this.  Why not separate this out so it can go in independently of that one?

asb committed:
Address review comment

asb commented:
> > Stacks on top of #120420.
> 
> I don't see any obvious dependency on this. Why not separate this out so it can go in independently of that one?

You're right, that would make more sense. As I don't have a good way of removing the dependency other than rebasing, I've removed the dependency and force pushed. I've updated the description to reflect this change.

preames commented on llvm/test/Transforms/DeadStoreElimination/memory-intrinsics-sizes.ll:
```diff
@@ -66,6 +66,71 @@ define void @memset_and_store_2(ptr %ptr, i64 %len) {
+
+define void @memset_pattern_different_size_values_1(ptr %ptr, i64 %len.1, i64 %len.2) {
+; CHECK-LABEL: @memset_pattern_different_size_values_1(
+; CHECK-NEXT:    call void @llvm.experimental.memset.pattern.p0.i8.i64(ptr align 1 [[PTR:%.*]], i8 0, i64 [[LEN_1:%.*]], i1 false)
```
A general enhancement we could do for this case would be to fold the calls to a length which is the umin of the two lengths.  This is generic and not related to this change.

preames commented on llvm/test/Transforms/DeadStoreElimination/memory-intrinsics-sizes.ll:
```diff
@@ -66,6 +66,71 @@ define void @memset_and_store_2(ptr %ptr, i64 %len) {
+  ret void
+}
+
+define void @memset_pattern_and_store_1(ptr %ptr, i64 %len) {
```
Can you add a positive test?  One which shows we *can* eliminate the store if the length is a constant?

preames commented on llvm/test/Transforms/DeadStoreElimination/memory-intrinsics-sizes.ll:
```diff
@@ -66,6 +66,71 @@ define void @memset_and_store_2(ptr %ptr, i64 %len) {
+  ret void
+}
+
+define void @memset_pattern_and_store_2(ptr %ptr, i64 %len) {
```
Possibly generic - but if we have a constant length, we could offset the start of the call and subtract 8 from length since we know the store follows.

preames reviewed (approved):
LGTM

asb committed:
Merge remote-tracking branch 'origin/main' into 2024q4-memset-pattern-memorylocation

asb committed:
Add test showing a store being eliminated

asb committed:
Merge remote-tracking branch 'origin/main' into 2024q4-memset-pattern-memorylocation
